### PR TITLE
SPM-1260: evaluate baseline

### PR DESCRIPTION
**Pull request by MichaelMraka**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: evaluator/evaluate_baseline.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	var filterOutNames []string
+	err = tx.Model(&models.AdvisoryMetadata{}).Where("name IN (?)", reportedNames).
+		Where("public_date >= ?", baseline.ToTime.Truncate(24*time.Hour)).
+		Pluck("name", &filterOutNames).Error
+	if err != nil {
+		return err
+	}
+
+	// create map of advisories we need to filter out
+	filterOutNamesMap := make(map[string]struct{}, len(filterOutNames))
```
Comment (Josca) [inline evaluator/evaluate_baseline.go:38]: It's actually set for us (not map), so I would name it filterOutNamesSet.

File: evaluator/evaluate_baseline.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	for pkg, updates := range vmaasData.GetUpdateList() {
+		availableUpdates := updates.GetAvailableUpdates()
+		filteredUpdates := make([]vmaas.UpdatesV2ResponseAvailableUpdates, 0, len(availableUpdates))
+		for _, u := range availableUpdates {
+			advisoryName := u.GetErratum()
+			if _, ok := filterOutNamesMap[advisoryName]; !ok {
+				filteredUpdates = append(filteredUpdates, u)
+			}
+		}
+		// update.SetAvailableUpdates(filteredUpdates)
```
Comment (Josca) [inline evaluator/evaluate_baseline.go:56]: Forgotten comment?

File: evaluator/evaluate_baseline_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	assert.Nil(t, err)
+	assert.Equal(t, expectedErrata, errataInVmaasData(originalVmaasData))
+
+	// a system with baseline and filtered errata
+	system = models.SystemPlatform{ID: 1, RhAccountID: 1, BaselineID: utils.PtrInt(1)}
+	vmaasData = getVMaaSUpdates(t)
+	expectedErrata = []string{"RH-100"}
+	err = limitVmaasToBaseline(database.Db, &system, &vmaasData)
+	assert.Nil(t, err)
+	assert.Equal(t, expectedErrata, errataInVmaasData(vmaasData))
```
Comment (Josca) [inline evaluator/evaluate_baseline_test.go:42]: Nice test. I would just use direct values in asserts. It's better understandable to me, and shorter. But not a big deal.
~~~golang
assert.Equal(t, []string{"RH-100"}, errataInVmaasData(vmaasData))
~~~

**Review by Josca — APPROVED** [review]
LGTM. Nice work! Just please update `clowdapp.yaml` and consider to use some cosmetic suggestions.

File: base/database/baseline.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+type BaselineConfig struct {
+	ToTime time.Time `json:"to_time"`
+}
+
+func GetBaselineConfig(tx *gorm.DB, system *models.SystemPlatform) (*BaselineConfig, error) {
+	var config BaselineConfig
+	var jsonb [][]byte
+
+	if system.BaselineID == nil {
```
[the remark this answers is not in the document]
Comment (MichaelMraka) [reply base/database/baseline.go:17]: fixed

File: evaluator/evaluate.go [hunk 5]
```diff
@@ -57,6 +58,7 @@ func configure() {
 	enableRepoAnalysis = utils.GetBoolEnvOrDefault("ENABLE_REPO_ANALYSIS", true)
 	enableStaleSysEval = utils.GetBoolEnvOrDefault("ENABLE_STALE_SYSTEM_EVALUATION", true)
 	enableLazyPackageSave = utils.GetBoolEnvOrDefault("ENABLE_LAZY_PACKAGE_SAVE", true)
+	enableBaselineEval = utils.GetBoolEnvOrDefault("ENABLE_BASELINE_EVAL", true)
```
[the remark this answers is not in the document]
Comment (MichaelMraka) [reply evaluator/evaluate.go:61]: Added. Currently I'll keep it disabled (false) until we have baseline API finished and can test it properly.

File: evaluator/evaluate_baseline.go (hunk 1 above)
Comment (MichaelMraka) [reply evaluator/evaluate_baseline.go]: ok

File: evaluator/evaluate_baseline.go (hunk 2 above)
Comment (MichaelMraka) [reply evaluator/evaluate_baseline.go]: removed comments

File: evaluator/evaluate_baseline_test.go (hunk 3 above)
Comment (MichaelMraka) [reply evaluator/evaluate_baseline_test.go]: fixed tests

**Comment by MichaelMraka** [comment]
Please don't merge it directly I'll squash fixups first after re-review.

**Review by Josca — APPROVED** [review]
:+1:

**Pull request merged**